packages/nimble-components/src/modal/index.ts

```diff
@@ -0,0 +1,123 @@
+import { attr } from '@microsoft/fast-element';
+import { FoundationElement } from '@microsoft/fast-foundation';
+import { UserDismissed } from '../patterns/dialog/types';
+
+/**
+ * This is a workaround for an incomplete definition of the native dialog element. Remove when using Typescript >=4.8.3.
+ * https://github.com/microsoft/TypeScript/issues/48267
+ * @internal
+ */
+export interface ExtendedDialog extends HTMLDialogElement {
+    showModal(): void;
+    close(): void;
+}
+
+export const ModalState = {
+    opening: 'opening',
+    open: 'open',
+    closing: 'closing',
+    closed: 'closed'
+} as const;
+export type ModalState =
+    (typeof ModalState)[keyof typeof ModalState];
+
+/**
+ * A base class for components based on a native dialog.
+ * @internal
+ */
+// eslint-disable-next-line @typescript-eslint/no-invalid-void-type
+export abstract class Modal<CloseReason = void> extends FoundationElement {
+    /**
+     * @public
+     * @description
+     * Prevents dismissing via the Escape key
+     */
+    @attr({ attribute: 'prevent-dismiss', mode: 'boolean' })
+    public preventDismiss = false;
+
+    /**
+     * The ref to the internal dialog element.
+     *
+     * @internal
+     */
+    public readonly dialogElement!: ExtendedDialog;
+
+    protected state: ModalState = ModalState.closed;
+
+    /**
+     * True if the Modal is open/showing, false otherwise
+     */
+    public get open(): boolean {
+        return this.state === ModalState.open;
+    }
+
+    private resolveShow?: (reason: CloseReason | UserDismissed) => void;
+
+    /**
+     * Opens the Modal
+     * @returns Promise that is resolved when the Modal is closed. The value of the resolved Promise is the reason value passed to the close() method, or UserDismissed if closed via the ESC key.
+     */
+    public async show(): Promise<CloseReason | UserDismissed> {
+        if (this.open) {
+            throw new Error('Already open');
+        }
+        this.startOpening();
+        return new Promise((resolve, _reject) => {
+            this.resolveShow = resolve;
+        });
+    }
+
+    /**
+     * Closes the Modal
+     * @param reason An optional value indicating how/why the Modal was closed.
+     */
+    public close(reason: CloseReason): void {
+        if (!this.open || this.state === ModalState.closing) {
```

Comment:
nit/pref: instead of accessing the public accessor inside the methods, which could have different meanings over time, lets compare the states instead. A nice syntax shortcut is something like `[ModalState.open, ModalState.closing].includes(this.state)`

Reply:
Actually, we don't even have to check if closing. The open check is sufficient.

Reply:
The type pattern we use for enums is to treat ModalState as a string. That's how we get type checked attribute strings in the angular templates, so the type `("open" | "closing")[]` is expected.

Looks like the surprising bit is the type for `includes` where I'd be on team the TypeScript is being too strict: https://github.com/Microsoft/TypeScript/issues/26255

So shortcut not worth the effort, just some good ol fashion `this.state === foo || this.state === bar`